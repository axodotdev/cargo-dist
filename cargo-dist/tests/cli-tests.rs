use std::process::{Command, Output, Stdio};

static BIN: &str = env!("CARGO_BIN_EXE_cargo-dist");

fn format_outputs(output: &Output) -> String {
    let stdout = std::str::from_utf8(&output.stdout).unwrap();
    let stderr = std::str::from_utf8(&output.stderr).unwrap();
    format!("stdout:\n{stdout}\nstderr:\n{stderr}")
}

#[test]
fn test_version() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("-V")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    let stdout = String::from_utf8(output.stdout).unwrap();
    let stderr = String::from_utf8(output.stderr).unwrap();

    assert!(output.status.success(), "{}", stderr);
    assert_eq!(stderr, "");

    let (name, ver) = stdout.split_once(' ').unwrap();
    assert_eq!(name, "cargo-dist");
    let mut ver_parts = ver.trim().split('.');
    ver_parts.next().unwrap().parse::<u8>().unwrap();
    ver_parts.next().unwrap().parse::<u8>().unwrap();
    let last = ver_parts.next().unwrap();
    if let Some((last, _prerelease)) = last.split_once('-') {
        last.parse::<u8>().unwrap();
    } else {
        last.parse::<u8>().unwrap();
    }

    assert!(ver_parts.next().is_none());
}

#[test]
fn test_long_help() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("--help")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("long-help", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_short_help() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("-h")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("short-help", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_manifest() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("manifest")
        .arg("--artifacts=all")
        .arg("--no-local-paths")
        .arg("--output-format=json")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    // We don't want this to churn every time we do a version bump
    insta::with_settings!({filters => vec![
        (r"\d+\.\d+\.\d+(\-prerelease\d+)?", "1.0.0-FAKEVERSION"),
        (r#""announcement_tag": .*"#, r#""announcement_tag": "CENSORED","#),
        (r#""announcement_title": .*"#, r#""announcement_title": "CENSORED""#),
        (r#""announcement_changelog": .*"#, r#""announcement_changelog": "CENSORED""#),
        (r#""announcement_github_body": .*"#, r#""announcement_github_body": "CENSORED""#),
        (r#""announcement_is_prerelease": .*"#, r#""announcement_is_prerelease": "CENSORED""#),
    ]}, {
        insta::assert_snapshot!(format_outputs(&output));
    });

    assert!(output.status.success(), "{}", output.status);
}
