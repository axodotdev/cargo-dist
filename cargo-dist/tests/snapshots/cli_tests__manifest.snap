---
source: cargo-dist/tests/cli-tests.rs
expression: format_outputs(&output)
---
stdout:
{
  "dist_version": "1.0.0-FAKEVERSION",
  "announcement_tag": "CENSORED",
  "announcement_is_prerelease": "CENSORED"
  "announcement_title": "CENSORED"
  "announcement_changelog": "CENSORED"
  "announcement_github_body": "CENSORED"
  "releases": [
    {
      "app_name": "cargo-dist",
      "app_version": "1.0.0-FAKEVERSION",
      "artifacts": [
        {
          "name": "cargo-dist-v1.0.0-FAKEVERSION-installer.sh",
          "kind": "installer",
          "target_triples": [
            "x86_64-apple-darwin",
            "x86_64-unknown-linux-gnu"
          ],
          "install_hint": "# WARNING: this installer is experimental\ncurl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v1.0.0-FAKEVERSION/cargo-dist-v1.0.0-FAKEVERSION-installer.sh | sh",
          "description": "Install prebuilt binaries via shell script"
        },
        {
          "name": "cargo-dist-v1.0.0-FAKEVERSION-installer.ps1",
          "kind": "installer",
          "target_triples": [
            "x86_64-pc-windows-msvc"
          ],
          "install_hint": "# WARNING: this installer is experimental\nirm https://github.com/axodotdev/cargo-dist/releases/download/v1.0.0-FAKEVERSION/cargo-dist-v1.0.0-FAKEVERSION-installer.ps1 | iex",
          "description": "Install prebuilt binaries via powershell script"
        },
        {
          "name": "cargo-dist-v1.0.0-FAKEVERSION-x86_64-apple-darwin.tar.xz",
          "kind": "executable-zip",
          "target_triples": [
            "x86_64-apple-darwin"
          ],
          "assets": [
            {
              "name": "LICENSE-APACHE",
              "path": "LICENSE-APACHE",
              "kind": "license"
            },
            {
              "name": "LICENSE-MIT",
              "path": "LICENSE-MIT",
              "kind": "license"
            },
            {
              "name": "README.md",
              "path": "README.md",
              "kind": "readme"
            },
            {
              "name": "RELEASES.md",
              "path": "RELEASES.md",
              "kind": "changelog"
            },
            {
              "name": "cargo-dist",
              "path": "cargo-dist",
              "kind": "executable"
            }
          ]
        },
        {
          "name": "cargo-dist-v1.0.0-FAKEVERSION-x86_64-pc-windows-msvc.zip",
          "kind": "executable-zip",
          "target_triples": [
            "x86_64-pc-windows-msvc"
          ],
          "assets": [
            {
              "name": "LICENSE-APACHE",
              "path": "LICENSE-APACHE",
              "kind": "license"
            },
            {
              "name": "LICENSE-MIT",
              "path": "LICENSE-MIT",
              "kind": "license"
            },
            {
              "name": "README.md",
              "path": "README.md",
              "kind": "readme"
            },
            {
              "name": "RELEASES.md",
              "path": "RELEASES.md",
              "kind": "changelog"
            },
            {
              "name": "cargo-dist",
              "path": "cargo-dist.exe",
              "kind": "executable"
            }
          ]
        },
        {
          "name": "cargo-dist-v1.0.0-FAKEVERSION-x86_64-unknown-linux-gnu.tar.xz",
          "kind": "executable-zip",
          "target_triples": [
            "x86_64-unknown-linux-gnu"
          ],
          "assets": [
            {
              "name": "LICENSE-APACHE",
              "path": "LICENSE-APACHE",
              "kind": "license"
            },
            {
              "name": "LICENSE-MIT",
              "path": "LICENSE-MIT",
              "kind": "license"
            },
            {
              "name": "README.md",
              "path": "README.md",
              "kind": "readme"
            },
            {
              "name": "RELEASES.md",
              "path": "RELEASES.md",
              "kind": "changelog"
            },
            {
              "name": "cargo-dist",
              "path": "cargo-dist",
              "kind": "executable"
            }
          ]
        }
      ]
    }
  ]
}

stderr:
analyzing workspace:
  cargo-dist
    [bin] cargo-dist
  cargo-dist-schema (no binaries)


